### PR TITLE
Topic/let patterns

### PR DESCRIPTION
**Pull request by HuwCampbell**

### CODE REVIEW

**Comment by HuwCampbell** [comment]
@amosr This is the thing I was discussing with you the other day (not really tidied up yet).

Essentially, it allows this sort of thing:

```haskell
feature age ~> let ((_, v), g) = (("that", value), "this") ~> sum v, newest g
```

Almost all of the "real" work is in the check and desugar phases.

**Comment by HuwCampbell** [comment]
I think if we continue down this road we should create a version of the AST which is simplified.

So like
```haskell
module Icicle.Source.Query.Context.Desugared where
data Context' q a n
 = Windowed a WindowUnit (Maybe WindowUnit)
 | Latest a Int
 | GroupBy   a             (Exp' q a n)
 | Filter    a             (Exp' q a n)
 | LetFold   a             (Fold q a n)
 | Let       a (Name n) (Exp' q a n)
 | GroupFold a (Name n) (Name n) (Exp' q a n)
 deriving (Show, Eq, Ord, Generic)

data Fold q a n
 = Fold
 { foldBind :: Name n
 , foldInit :: Exp' q a n
 , foldWork :: Exp' q a n }
 deriving (Show, Eq, Ord, Generic)
```

Essentially, all parts of the source AST which can't be converted to core
aren't possible in the desugared AST.

File: icicle-source/src/Icicle/Source/ToCore/Exp.hs [hunk 1]
```diff
@@ -108,17 +107,36 @@ convertExpQ q
...
+    (Let _ PatDefault _:cs)
+     ->     -- A pattern with no binding doesn't need to inspect its arguments
+            convertExpQ $ Query cs $ final q
+    (Let _ (PatVariable b) d:cs)
      -> do  d' <- convertExp d
             -- NB: because it's non-recursive let, the freshen must be done after the definition
             b' <- convertFreshenAdd b
             x' <- convertExpQ $ Query cs $ final q
             return $ CE.XLet () b' d' x'
+    -- (Let ann (PatCon ConTuple [PatVariable na, PatVariable nb]) bound:cs)
```
Comment (amosr) [inline icicle-source/src/Icicle/Source/ToCore/Exp.hs]: remnant cruft?

**Comment by amosr** [comment]
👍 looks good

> I think if we continue down this road we should create a version of the AST which is simplified.

yep, agree. I think the ideal would be to have Desugar go from Sorbet to an explicitly typed Source with explicit box/unbox calls, but that's a bigger change for not much benefit right now

File: icicle-source/src/Icicle/Source/ToCore/Exp.hs (hunk 1 above)
Comment (HuwCampbell) [reply icicle-source/src/Icicle/Source/ToCore/Exp.hs]: Yeah, I tried to do the translation here first, but it was a bit subtle. I had it working for `let`, but then some possibilities code was breaking.

**Comment by HuwCampbell** [comment]
I'm kind of in two minds. There are two conceptual paths to Sorbet I think.

1. Enhance Source as it stands to be as powerful as Sorbet, then the Sorbet elaborator is essentially just going from CST to AST. This is what this work would be a part of.

2. More or less what you've said, leave Source as is or cut it down to something slightly neater, and desugar sorbet to that.

The benefits of the first approach is that type checking would be down _closer_ to the source code the user has written, so it might give better error messages. Downside is that there should be another desugared AST as well.

Benefits of the second approach is a simpler type checker, and less impact on the current AST.

**Comment by amosr** [comment]
yes, sorry, I really meant closer to 1, but I left out a whole bunch of context that was only in my head and didn't explain it properly. when I said Sorbet, I really meant Source which has been enhanced with the features of Sorbet. I forgot we already had a datatype for Sorbet sitting around.

your Desugared AST looks good. the gap between Source and Core is already a bit too large, so having Desugared would help. what I really meant to say was that we could add extra type information to the Desugared, because it might make conversion easier. as it is, the type annotations are about the return type, so the Let case:

```
Let a (Name n) (Exp' q a n)
```

when the `a` includes type annotations, it only says what the return type of the overall expression is, nothing about the binding itself. you can look at the annotation for the expression to find the bound type, but to figure out how to translate it you need to look at the modality of the bound and the modality of the remaining query and context. this works, but I think it would be nicer overall if we made this information explicit in the Desugared AST. we shouldn't do this in the first stage, but if we decide to later, it will be easier with your Desugared AST.

as an aside, I feel like there's some sweet spot in between Source and Core that we're missing. filter and group shouldn't be primitives, etc.

**Comment by HuwCampbell** [comment]
```haskell
λ feature injury ~> fold (count, sum) = (0, 0) : (count + 1, sum + severity) ~> count, sum
C evaluation
------------

homer|(5, 20.0)
marge|(5, 20.0)
moe|(2, 3.0)
```

**Comment by HuwCampbell** [comment]
This all seems to work pretty well, but I have a small issue on let bindings or many pure values.
```
let (a, b) = (1, 2)
```
doesn't seem to work too well.

**Pull request merged**